### ISSUE TITLE
PlayJava version of library?

### ISSUE THREAD

**Issue by achaiah** [adrianhurt/play-bootstrap#50]
Would you be able to make a copy of this library that compiles with "enablePlugins(PlayJava)" rather than "enablePlugins(PlayScala)"?  I tried it myself but I don't have enough experience with sbt to resolve all the compilation issues.

This would be very helpful to people who are creating Play java projects but are using a different build tool (e.g. gradle). Otherwise, there are compilation issues when including your library with a Java play project.

**Comment by adrianhurt** [comment]
Hi @achaiah. Sorry but I don't understand what you exactly need. The library should work in a Play Java project (with `enablePlugins(PlayJava)`) without problems importing it with:
```
libraryDependencies ++= Seq(
 …,
  "com.adrianhurt" %% "play-bootstrap" % "1.0-P24-B3-SNAPSHOT"
)
```

Do you need a version to integrate the code directly within your project and without `libraryDependencies`? I only work with Scala so I'm not very sure, but I think Play understands Java and Scala files independently. If you need that, you can try just copy and paste the code and try it, or try to port the Scala files to Java (there are a very few code that it's not within the templates). But sorry, I can't port the library to Java. I don't use Java Play since Play 1 and I don't have time to do that. Anyway, if you have specific errors, you can put them here and I could try to help you.

**Comment by achaiah** [comment]
Hi,

Thanks for your reply. I'm trying to use play with gradle to simplify the building process and it mostly works fine. Where I run into issues is that play-bootstrap uses play.api.i18n.Messages in core-play24/app/views/bs/package.scala and play24-bootstrap3/... and unfortunately that creates compilation issues when trying to use play-bootstrap inside a java project (when used with gradle).

I actually converted most of the conflicting code to java already. The only thing I cannot figure out is how I could use play.i18n.Messages instead of play.api.i18n.Messages in your code. I'd be grateful if you could help me in any way.

**Comment by adrianhurt** [comment]
I don't really know… I have seen this https://www.playframework.com/documentation/2.4.x/JavaI18N#Use-in-templates. Maybe you can use the Scala Messages in your code...

**Comment by adrianhurt** [comment]
Looking a bit the docs of gradle… Can't you adding the library as a normal dependency? Look at section "71.8.2. Adding dependencies" here https://docs.gradle.org/current/userguide/play_plugin.html. Maybe you can add it with:
```
dependencies {
    play "com.adrianhurt:play-bootstrap:1.0-P24-B3-SNAPSHOT"
}
```

**Comment by achaiah** [comment]
Yup, that's exactly what I did -- I added your library as a dependency. The trouble comes when the play framework forms are compiled. The elements in your project (e.g. `play24/app/views/bs/package.scala`) expect a `play.api.i18n.Messages` object but when a Java-based view is generated (e.g. when one calls `return ok(views.html.profile.render(formName))`) a `play.i18n.Messages` is passed into the form.  While you don't explicitly declare dependency on the `Messages` class, it is implicitly passed into your form and it seems that it's extensively used in `play24/app/views/bs/package.scala`.

Unfortunately, as far as I know, there is no way for me to coerce the compiler to use a `play.api.i18n.Messages` instead because it uses Java classes. Interestingly, there are other implicit differences between Java and Scala versions of play that I had to work around but I solved every single one except the `Messages`.  Sadly, I'm stuck since they seem to play an important role in the code and I'm not a scala developer :(

**Comment by adrianhurt** [comment]
I have this within the "Multifield" and "Extend it!" sections in the docs. Have you tried it?

> __Note for Java version from Play 2.4__
From Play 2.4, the implicit Messages argument defined in your custom template collides with the implicit one taken from PlayMagicForJava object, and it raises a compilation error. So you only need to remove this implicit argument from your custom template.

**Comment by achaiah** [comment]
Ok that's interesting. The forms I'm trying to compile are from this seed: [PlayAuthenticate with Morphia](https://github.com/esfand-r/Play2.3-Spring-PlayAuthenticate-deadbolt2-and-mongo-with-morphia).  So looking at this [Password form](https://github.com/esfand-r/Play2.3-Spring-PlayAuthenticate-deadbolt2-and-mongo-with-morphia/blob/master/modules/web/app/views/_passwordPartial.scala.html), for example, I can see that it uses `@import b3.vertical.fieldConstructor` and there are no implicits declared anywhere. Is this what you mean?  If so then it's not enough as the b3.password looks to be defined like this in your documentation: `b3.password @(field: Field, args: (Symbol,Any)*)(implicit handler: B3FieldConstructor, messages: Messages)`

I think I may be able to get the correct Scala Messages object by using PlayMagicForJava but I don't know how I would call your `b3.password` constructor with it.  Given this [Password form](https://github.com/esfand-r/Play2.3-Spring-PlayAuthenticate-deadbolt2-and-mongo-with-morphia/blob/master/modules/web/app/views/_passwordPartial.scala.html), how could I pass in a Scala Messages object, if I could get one via PlayMagicForJava?

**Comment by adrianhurt** [comment]
Ok, so I understand now this is a problem with Play Java and not for gradle, isn't it? As I told you I don't really know how Play Java works, but as far as I know, every template in Play Java takes a `play.api.i18n.Messages` thanks to [PlayMagicForJava. implicitJavaMessages](https://github.com/playframework/playframework/blob/master/framework/src/play-java/src/main/scala/play/core/TemplateMagicForJava.scala). That's why this __password form__ example doesn't need to specify explicitly any `B3FieldConstructor` nor `Messages`; they are implicitly taken from `@import b3.vertical.fieldConstructor` and `PlayMagicForJava. implicitJavaMessages`. So you can assume the constructor of `b3.password` is simply `(field: Field, args: (Symbol,Any)*)`.

But if you need to have a reference of the corresponding `play.api.i18n.Messages` for any reason. You could get it with something like (I haven't test it …):
```
@define(implicitly[play.api.i18n.Messages]) { msgs =>
   … use here your Scala Messages …
  @b3.password(f("password"), '_label -> "")(implicitly[b3.B3FieldConstructor], msgs)
}
```
[Check this to know a bit more](http://adrianhurt.github.io/play-bootstrap/1.0-P24-B3/mixed/#mixed-within-form).

But the magic of implicits in Scala is that you can omit them when you know they are there.

**Comment by achaiah** [comment]
Well, the issue with gradle is that while I cannot compile it with gradle, I can compile it with sbt when I specify `enablePlugins(PlayJava)` in the build script. There is something "magical" that got broken in Play 2.4 release that doesn't allow gradle to perform the same magic as the sbt plugin.

But... it seems like we've almost found a work-around! So in that Password Form I now have a declaration like this:

```
@defining(implicitly[play.api.i18n.Messages]) { msgs =>
        PlayMagicForJava.implicitJavaMessages
    @b3.password(PlayMagicForJava.javaFieldtoScalaField(f("password")), '_label -> "")(implicitly[b3.B3FieldConstructor], msgs)
    }
```

But that doesn't seem to compile. I get:

```
could not find implicit value for parameter e: play.api.i18n.Messages
    """),_display_(/*15.6*/defining(implicitly[play.api.i18n.Messages])/*15.50*/ { msgs =>_display_(Seq[Any](format.raw/*15.60*/("""
                                              ^

```

It's not happy about that "implicitly" block... maybe I'm a bit off on the syntax?

**Comment by achaiah** [comment]
Oh, I think I got it...or at least it compiles now :)

`@b3.password(PlayMagicForJava.javaFieldtoScalaField(f("password")), '_label -> "")(implicitly[b3.B3FieldConstructor], PlayMagicForJava.implicitJavaMessages)`

**Comment by adrianhurt** [comment]
Ok, then try this:
```
@(f: Form[_])
@import b3.vertical.fieldConstructor
@defining(PlayMagicForJava.implicitJavaMessages) { implicit msgs =>
   @b3.password(PlayMagicForJava.javaFieldtoScalaField(f("password")), '_label -> "")
}
```

Then, you could simplify your syntax for every helper declaration.

Is it necessary to add `PlayMagicForJava.javaFieldtoScalaField`? If so, maybe you can import it in every template (`@import PlayMagicForJava.javaFieldtoScalaField `).

Anyway, then I think this is a `PlayMagicForJava` issue. It should be imported automatically but it's not. So now that you have localized it, maybe you can google a bit.

Maybe you can also try to import it directly within your templates `@import PlayMagicForJava._` or something like that. I know there is a way to declare import automatically for every template using sbt.

**Comment by adrianhurt** [comment]
or better, maybe something like this could work for you:
```
@(f: Form[_])
@import PlayMagicForJava.javaFieldtoScalaField
@implicitMessages = @{PlayMagicForJava.implicitJavaMessages}
@import b3.vertical.fieldConstructor

@b3.password(f("password"), '_label -> "")
```

**Comment by achaiah** [comment]
Hmm interesting, I'll try that!

**Comment by achaiah** [comment]
So I've played around with it for a bit. If I just use the import statements it compiles in gradle but not in sbt (sbt complains about ambiguous implicit input). But something like this:

`@b3.password(PlayMagicForJava.javaFieldtoScalaField(f("password")), '_label -> "")(implicitly[b3.B3FieldConstructor], PlayMagicForJava.implicitJavaMessages)`

will compile in both.  Looks like something got changed in play 2.4.x that causes PlayMagicForJava to break when not using sbt. It's a shame really.

Thanks a lot for your help!
